core/src/main/scala/caliban/rendering/Renderer.scala

```diff
@@ -0,0 +1,559 @@
+package caliban.rendering
+
+import caliban.{ InputValue, Rendering }
+import caliban.parsing.adt.Definition.ExecutableDefinition.{ FragmentDefinition, OperationDefinition }
+import caliban.parsing.adt.Definition.TypeSystemDefinition.TypeDefinition.{
+  EnumTypeDefinition,
+  EnumValueDefinition,
+  FieldDefinition,
+  InputObjectTypeDefinition,
+  InputValueDefinition,
+  InterfaceTypeDefinition,
+  ObjectTypeDefinition,
+  ScalarTypeDefinition,
+  UnionTypeDefinition
+}
+import caliban.parsing.adt.Definition.TypeSystemDefinition.{ SchemaDefinition, TypeDefinition }
+import caliban.parsing.adt.Type.{ innerType, NamedType }
+import caliban.parsing.adt.{ Definition, Directive, Document, OperationType, Selection, VariableDefinition }
+
+trait Renderer[-A] { self =>
+
+  def contramap[B](f: B => A): Renderer[B] = new Renderer[B] {
+    override def unsafeRender(value: B, indent: Option[Int], write: StringBuilder): Unit =
+      self.unsafeRender(f(value), indent, write)
+  }
+
+  def optional: Renderer[Option[A]] = new Renderer[Option[A]] {
+    override def unsafeRender(value: Option[A], indent: Option[Int], write: StringBuilder): Unit =
+      value.foreach(self.unsafeRender(_, indent, write))
+  }
+
+  def list: Renderer[List[A]] = new Renderer[List[A]] {
+    override def unsafeRender(value: List[A], indent: Option[Int], write: StringBuilder): Unit =
+      value.foreach(self.unsafeRender(_, indent, write))
+  }
+
+  def list(separator: String): Renderer[List[A]] = new Renderer[List[A]] {
+    override def unsafeRender(value: List[A], indent: Option[Int], write: StringBuilder): Unit = {
+      var first = true
+      value.foreach { v =>
+        if (first) first = false
+        else write.append(separator)
+        self.unsafeRender(v, indent, write)
+      }
+    }
+  }
+
+  def unsafeRender(value: A, indent: Option[Int], write: StringBuilder): Unit
+}
+
+object Renderer {
+
+  private implicit val typeOrdering: Ordering[TypeDefinition] = Ordering.by {
+    case TypeDefinition.ScalarTypeDefinition(_, name, _)          => (0, name)
+    case TypeDefinition.UnionTypeDefinition(_, name, _, _)        => (1, name)
+    case TypeDefinition.EnumTypeDefinition(_, name, _, _)         => (2, name)
+    case TypeDefinition.InputObjectTypeDefinition(_, name, _, _)  => (3, name)
+    case TypeDefinition.InterfaceTypeDefinition(_, name, _, _, _) => (4, name)
+    case TypeDefinition.ObjectTypeDefinition(_, name, _, _, _)    => (5, name)
+  }
+
+  def render(document: Document): String = {
+    val builder = new StringBuilder
```

Comment:
The default StringBuilder has a size of 16, which causes it to keep resizing once we reach that size. Given that these rendered strings are likely to be more than 16 chars, is there a more sensible default we could use? Perhaps 32 or 64?

Reply:
Yeah I think presetting it to the size of `SourceMapper` (or rounding up to the next power of 2?) makes sense since the rendered document is likely going to be smaller than that? Unless the query is minimized of course.

Reply:
Done (I think)